Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp

```diff
@@ -436,7 +436,16 @@ WebIDL::ExceptionOr<void> XMLHttpRequest::send(Optional<Fetch::XMLHttpRequestBod
     if (m_method.is_one_of("GET"sv, "HEAD"sv))
         body = {};
 
-    auto body_with_type = body.has_value() ? TRY(Fetch::extract_body(realm, body.value())) : Optional<Fetch::Infrastructure::BodyWithType> {};
+    Optional<Fetch::Infrastructure::BodyWithType> body_with_type {};
+    Optional<String> serialized_document {};
+    if (body.has_value()) {
+        if (body->has<JS::Handle<DOM::Document>>()) {
+            serialized_document = TRY(body->get<JS::Handle<DOM::Document>>().cell()->serialize_fragment(DOMParsing::RequireWellFormed::No));
+        } else {
+            auto body_init_or_readable_bytes = body->downcast<JS::Handle<Streams::ReadableStream>, JS::Handle<FileAPI::Blob>, JS::Handle<JS::Object>, JS::Handle<URL::URLSearchParams>, String, ReadonlyBytes>();
+            body_with_type = TRY(Fetch::extract_body(realm, body_init_or_readable_bytes));
```

Comment:
CxByte kindly added this in https://github.com/SerenityOS/serenity/pull/16002:

```suggestion
            body_with_type = TRY(Fetch::extract_body(realm, body->downcast<BodyInitOrReadableBytes>()));
```

Reply:
So I can actually downcast it to a completely different Variant, CxByte really is a wizard